### PR TITLE
Additional detailed description for window impls

### PR DESCRIPTION
**Pull request by mohanvive**

### CODE REVIEW

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 1]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+* timeAccum
+* hopping
+* timeOrder
+
+
+1.  time window
+
+    `time(int windowTime)`
+
+    A sliding time window that holds events that arrived during the last `windowTime` period at a given time, and
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    At a given time, this sliding time window holds the events that arrived during the last `windowTime` period. 
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 2]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    gets updated for each event arrival and expiry.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    Also, it gets updated on the arrival and expiry of each event.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 3]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+2.  timeBatch window
+
+    `timeBatch(int windowTime)`
+
+    A batch (tumbling) time window that holds events that arrive during `windowTime` periods, and gets updated for
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A batch (tumbling) time window holds the events that arrive during `windowTime` periods, and gets updated on
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 4]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    each `windowTime`.
+
+3. timeLength window
+
+    `timelength(int windowTime, int windowLength)`
+
+    A sliding time window that, at a given time holds the last `windowLength` events that arrived during last
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    At a given time, a sliding time window holds the last `windowLength` events that arrived during the last
```
Is it the "the last `windowLength` event or events?

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 5]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    `windowTime` period, and gets updated for every event arrival and expiry.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    `windowTime` period, and gets updated on the arrival and expiry of every event.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 6]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+4. length window
+    `length(int windowLength)`
+
+    A sliding length window that holds the last `windowLength` events at a given time, and gets updated for each
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    At a given time, a sliding length window holds the last `windowLength` events, and gets updated on 
```
The last `windowLength` event or events?

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 7]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    arrival and expiry.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
   the arrival and expiry of each event.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md (hunk 2 above)
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 8]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+    A sliding time window that holds events that arrived during the last `windowTime` period at a given time, and
+    gets updated for each event arrival and expiry.
+
+2.  timeBatch window
+
+    `timeBatch(int windowTime)`
+
+    A batch (tumbling) time window that holds events that arrive during `windowTime` periods, and gets updated for
+    each `windowTime`.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion

```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 9]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    `length(int windowLength)`
+
+    A sliding length window that holds the last `windowLength` events at a given time, and gets updated for each
+    arrival and expiry.
+
+5. lengthBatch window
+
+    `lengthBatch(int windowLength)`
+
+    A batch (tumbling) length window that holds a number of events specified as the `windowLength`. The window is
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A batch (tumbling) length window holds a number of events of a specific`windowLength`. The window is
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 10]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    updated each time a batch of events that equals the number specified as the `windowLength` arrives.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    updated each time a batch of events, equal to the number specified as the `windowLength` arrives.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 11]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+6. externalTime window
+
+    `externalTime(timeStamp, int windowTime)`
+
+    A sliding time window based on external time. It holds events that arrived during the last `windowTime` period
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A sliding time window is based on external time. It holds events that arrived during the last `windowTime` period
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 12]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    from the external `timestamp`, and gets updated on every monotonically increasing `timestamp`. Here the
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    from the external `timestamp`, and gets updated on every `timestamp` that increases monotonically. The
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 13]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    `timeStamp` should be an attribute of the record which is used as the constraint type of relevant input stream.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    `timeStamp` should be an attribute of the record, which is used as the constraint type of the relevant input stream.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 14]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    As the `timeStamp` parameter you should pass `<streamName>.<attiributeName>`.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    You should pass `<streamName>.<attiributeName>`as the `timeStamp` parameter.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 15]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+7. externalTimeBatch window
+
+    `externalTimeBatch(timeStamp, int windowTime, int? startTime, int? timeout)`
+
+    A batch (tumbling) time window based on external time, that holds events arrived during `windowTime` periods, and
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A batch (tumbling) time window is based on external time. It gets updated every `windowTime` and holds events that arrived during the specified `windowTime` periods.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 16]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+     gets updated for every `windowTime`. Here the `timeStamp` should be an attribute of the record which is used as
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
     The `timeStamp` should be an attribute of the record, which is used as
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 17]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+     the constraint type of relevant input stream. As the `timeStamp` parameter you should pass `<streamName>
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
     the constraint type of relevant input stream. You should pass `<streamName>` as the value of the`timeStamp` parameter. 
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 18]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+     .<attiributeName>`. Parameters `startTime` and `timeout` are optional parameters. `startTime` can be used to
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
     The `startTime` and `timeout` parameters are optional. `startTime` can be used to
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 19]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+     specify a user defined time to start the first batch. `timeout` is time to wait for arrival of new event, before
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
     specify a time defined by the user to start the first batch. `timeout` is the time to wait for the arrival of a new event, before
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 20]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+      flushing and giving output for events belonging to a specific batch. Usually `timeout` is greater than
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
      flushing and sending output for events belonging to a specific batch. Usually, `timeout` is greater than
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 21]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+      `windowTime`.
+
+8. uniqueLength window
+
+    `uniqueLength(uniqueAttribute, int windowLength)`
+
+    A sliding length window that returns unique events within the `windowLength` based on the given `uniqueAttribute`
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]: A sliding length window that returns unique events within the `windowLength` based on the given `uniqueAttribute`. The `uniqueAttribute` should be an attribute of the record, which is used as the constraint type of the relevant input stream.

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 22]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    . Here the `uniqueAttribute` should be an attribute of the record which is used as the constraint type of
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 23]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    relevant input stream.
+
+9. delay window
+
+    `delay(int delayTime)`
+
+    A delay window holds events for a specific time period(`delayTime`) that is regarded as a delay period before
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A delay window holds events of a specific time period(`delayTime`), which is regarded as a delay period before
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 24]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    processing them.
+
+10. sort window
+
+    `sort(int windowLength, attributeName, string order)`
+
+    A sort sort window holds a batch of events that equal the number specified as the `windowLength` and sorts them
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]: A sort window holds a batch of events that equals the number specified as the `windowLength` and sorts them in the given `order` of the given `attributeName`. The `attributeName` should be an attribute of the record, which is used as the constraint type of the relevant input stream. You can have multiple `attributeName` fields followed by its `order`.

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 25]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    in the given `order` of given `attributeName`. Here the `attributeName` should be an attribute of the record
+    which is used as the constraint type of relevant input stream. You can have multiple `attributeName` fields
+    followed by it's `order`.
+
+11. timeAccum window
+
+    `timeAccum(int timePeriod)`
+
+    A sliding window that accumulates events until no more events arrive within the `timePeriod`, and only then
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A sliding window that accumulates events until no more events arrive within the `timePeriod`, and releases them only then.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 26]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+    releases the accumulated events.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
   
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 27]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+12. hopping window
+
+    `hopping(int windowTime, int hoppingTime)`
+
+    A hopping window holds the events arrived within last `windowTime` and release them in every `hoppingTime` period.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]:
```suggestion
    A hopping window holds the events that arrived within the last `windowTime` and releases them in every `hoppingTime` period.
```

File: website/two-column-pages/docs/learn/ballerina-streaming-reference.md [hunk 28]
```diff
@@ -463,11 +463,117 @@ Following are some inbuilt windows shipped with Ballerina Streams.
...
+
+13. timeOrder window
+
+    `timeOrder(timestamp, int windowTime, boolean dropOlderEvents)`
+
+    A timeOrder window orders events that arrive out-of-order, using timestamp values provided by `timestamp`, and
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/ballerina-streaming-reference.md]: A timeOrder window orders events that arrive out-of-order by comparing  the `timestamp` values against the system time. `windowTime` is the duration of the window. 

The `dropOlderEvents` flag determines whether to drop the events, which have timestamp value less than the tail-time of the current window. Tail-time is a period of `windowTime` before the system time. 

The `timeStamp` should be an attribute of the record, which is used as the constraint type of the relevant input stream. You should pass `<streamName>.<attiributeName>` as the values of the `timeStamp` parameter.

**Comment by keizer619** [comment]
@praneesha  @mohanvive 
We will be merging this PR to get website update and lets send another PR by fixing above suggestions

**Comment by praneesha** [comment]
The descriptions of the window definitions are reviewed in https://github.com/ballerina-platform/ballerina-www/pull/1388.

**Pull request merged**